### ISSUE TITLE
Language environments

### ISSUE THREAD

**Issue by andreaferretti** [tailhook/vagga#171]
It seems that Vagga is aware of Pip and Npm. Is there a way to make commands for other environments or are these things hard-coded?

For Python and Node: are the versions hard-coded? For instance, it would be nice to use Conda instead of Pip for Python, or use Nvm to switch Node versions

**Comment by tailhook** [comment]
Yes, the commands are hardcoded. This makes everybody unhappy at the start, but it's not that bad as it lools like:

First, can mostly do the same with shell commands (`!Sh`) and few helpers (like  !CacheDir). If you have specific task.

Second, the amount of "hard code" is not that much:
```
$ wc -l src/builder/commands/{npm,pip}.rs
  165 src/builder/commands/npm.rs
  182 src/builder/commands/pip.rs
```

> For Python and Node: are the versions hard-coded? 

You can disable installation of python using `install-python: false` and specify `python-exe` in [PipConfig](http://vagga.readthedocs.org/en/latest/build_steps.html#step-PipConfig). Currently, you can't disable pip, but feel free to submit a pull request

Simiarly, in nodejs there is `install-node: false` and `npm-exe`.

So yes, the idea is to make it easy to build containers.While you can do anything using plain shell commands like you do with docker, I think that 80% of the tasks can be done using convenience commands.

Feel free to contribute more commands and settings.

**Comment by andreaferretti** [comment]
I understand. It would be nice if the mechanism was made expandible via some sort of plugin, so that one could have the same integration for Npm and Python also for other languages, such as Scala, Nim, Rust... I understand if anyway this is not a priority right now

**Comment by tailhook** [comment]
@andreaferretti could you elaborate on what kind of expansion would you like?

The idea is that ad-hoc expansion, just for one project, works by using shell commands.

When you want to provide a new language environment, you are going to write rust code anyway.

So there are no plugins for several reasons:

1. I want every vagga file work with every vagga installation because it's vagga's job to remove dependency hell. Requiring plugins for specific `vagga.yaml` would be dependency hell again
2. Having project-specific dependencies is a security issue. With current design, you can run untrusted code with vagga and be sure that it never touches the system
3. Rust have no stable ABI, so we can't make binary plugins. At least not in a way that it's convenient to write.

**Comment by andreaferretti** [comment]
Yeah, you are right that having plugins would introduce a bunch of downsides. On the other hand, it seems just weird that Python and Node are given special status in the project.

After a little bit of thinking, I decided that I would not use this feature anyway. It seems more convenient to use [nvm](https://github.com/creationix/nvm) plus `package.json` for Node, or [miniconda](http://conda.pydata.org/miniconda.html) plus `setup.py` for Python in order to set up dependencies.

In other words: wherever a sensible language-specific package manager is available, I would rather use Vagga to configure the environment and possibly some binary DLLs, and then use the package manager for the dependencies themselves.

I know that NPM and SetupTools are limited, but they would be needed anyway, since in production there should be some way to retrieve the dependencies.

**Comment by tailhook** [comment]
Well, now that we have also PHP and Ruby, should this be closed?

**Comment by andreaferretti** [comment]
Great to know! I am closing the issue :-)
